File: communicator/mqtt_topic.py
from abc import ABC

class MqttTopic(ABC):
    SINGLELEVEL_WILDCARD = "+"
    MULTILEVEL_WILDCARD = "#"
    DELIMITER = "/"
    
    def __init__(self, topic_string=None):
        if topic_string:
            self.tokens = topic_string.split("/")
        else:
            self.tokens = None
    
    def __str__(self):
        if not self.tokens:
            return ""
            
        result = []
        for token in self.tokens:
            if token is not None:
                result.append(token)
            else:
                result.append(self.SINGLELEVEL_WILDCARD)
        
        return self.DELIMITER.join(result)

File: lapras-middleware/src/main/java/kr/ac/kaist/cdsn/lapras/communicator/MqttTopic.java
package kr.ac.kaist.cdsn.lapras.communicator;

/**
 * Created by Daekeun Lee on 2016-11-15.
 */
public abstract class MqttTopic {
    public static String SINGLELEVEL_WILDCARD = "+";
    public static String MULTILEVEL_WILDCARD = "#";
    public static String DELIMITER = "/";

    protected String[] tokens;

    public MqttTopic(String topicString) {
        tokens = topicString.split("/");
    }

    protected MqttTopic() {
    }

    public String toString() {
        StringBuilder sb = new StringBuilder();
        for(int i = 0;i<tokens.length;i++) {
            String token = tokens[i];
            if(token != null) {
                sb.append(token);
            } else {
                sb.append(SINGLELEVEL_WILDCARD);
            }
            if(i + 1 < tokens.length) {
                sb.append(DELIMITER);
            }
        }
        return sb.toString();
    }
}
